src/PoolManager.sol

```diff
@@ -207,34 +218,71 @@ contract PoolManager is IPoolManager, Fees, NoDelegateCall, ERC6909Claims {
-                tickSpacing: key.tickSpacing,
-                zeroForOne: params.zeroForOne,
-                amountSpecified: params.amountSpecified,
-                sqrtPriceLimitX96: params.sqrtPriceLimitX96
-            })
+        // The hook's deltas are from the point of view of the hook. Positive: the hook took money, negative: the hook sent money to the pool
+        // TODO consider adding hookDeltaInUnspecified to beforeSwap hook too
+        int128 hookDeltaInSpecified = key.hooks.beforeSwap(key, params, hookData);
+
+        bool exactInput = params.amountSpecified > 0;
+        params.amountSpecified -= hookDeltaInSpecified;
+        require(exactInput ? params.amountSpecified > 0 : params.amountSpecified < 0);
+
+        // ExactIn (amountSpecified is positive)
+        // In the case where the hook contributed x, x is given to the user now. After the swap amountSpecified + x input are taken from the user, meaning the user pays amountSpecified overall.
+        // In the case where the hook took x, x is taken from the user now. After the swap amountSpecified - x input are taken from the user, meaning the user pays amountSpecified overall.
+        // ExactOut (amountSpecified is negative)
+        // In the case where the hook contributed x, x output is given to the user now. The swap of amountSpecified + x is swapping for x fewer output tokens - so the user gets amountSpecified overall.
+        // In the case where the hook took x, x output is taken from the user now. The swap of amountSpecified - x is swapping for x extra output tokens - so the user gets amountSpecified overall.
+        _accountDeltaFor(
+            (exactInput == params.zeroForOne) ? key.currency0 : key.currency1, hookDeltaInSpecified, address(key.hooks)
         );
 
-        _accountPoolBalanceDelta(key, delta);
-
-        // the fee is on the input currency
-        unchecked {
-            if (feeForProtocol > 0) {
-                protocolFeesAccrued[params.zeroForOne ? key.currency0 : key.currency1] += feeForProtocol;
+        {
+            uint256 feeForProtocol;
+            uint24 swapFee;
+            Pool.SwapState memory state;
+            (delta, feeForProtocol, swapFee, state) = pools[id].swap(
+                Pool.SwapParams({
+                    tickSpacing: key.tickSpacing,
+                    zeroForOne: params.zeroForOne,
+                    amountSpecified: params.amountSpecified,
+                    sqrtPriceLimitX96: params.sqrtPriceLimitX96
+                })
+            );
+
+            // the fee is on the input currency
+            unchecked {
+                if (feeForProtocol > 0) {
+                    protocolFeesAccrued[params.zeroForOne ? key.currency0 : key.currency1] += feeForProtocol;
+                }
             }
+
+            // TODO returning deltas here doesnt make any sense
+            emit Swap(
+                id,
+                msg.sender,
+                -delta.amount0(),
```

Comment:
and why flip them here?

Reply:
this is a remnant of flipping deltas, fixed in emily's flip of amountSpecified - itll be removed shortly when i merge that update in here 🙌